### PR TITLE
[ARRAY]Save array's size instead of offset.

### LINKED ISSUE #5142: Optimize nested type implementation
Array is used to store data like user labels.
made up of (offset, element), element can be an Array.
[1, 2, 3], [4, 5, 6]
Doris have implement array nest type. But it's not extensible.
```
1. It store NULL with offset, which only be feasible to Array, but not to Struct.
    Struct is made of {element 1, element2}, but not has implicit offset column.
    Array should store null as a new column, not attaches to offset column.
2. It stores offset with absolute ordinal, so every offset ColumnWriter will have to 
    callback ArrayColumnWriter to record the position. The logic is so trick to understand.
3. When read the ArrayColumn, it have to seek to next position to get the right end offset.
```

### PR DESCRIPTION
**Pull request by liutang123**
## Proposed changes

In #5142 , chaoyli gives some suggestions:

1. ArrayColumnWriter writes null as separate column.
2. save array's size instead of child column's offset.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x]  chaoyli have created an issue on (Fix #5142) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

### CODE REVIEW

**Comment by liutang123** [comment]
@yangzhg cc @wangbo

File: be/test/olap/column_vector_test.cpp [hunk 1]
```diff
@@ -72,60 +72,70 @@ void test_read_write_scalar_column_vector(const TypeInfo* type_info, const uint8
...
+    ColumnVectorBatch* item_cvb = array_cvb->elements();
+    ColumnVectorBatch* offset_cvb = array_cvb->offsets();
 
     // first write
-    list_cvb->put_item_ordinal(ordinals, 0, list_init_size + 1);
-    list_cvb->set_null_bits(0, list_init_size, false);
-    size_t first_write_item = ordinals[list_init_size] - ordinals[0];
+    for (size_t i = 0; i < array_init_size; ++i) {
+        memcpy(offset_cvb->mutable_cell_ptr(1 + i), &(result[i].length), sizeof(segment_v2::ordinal_t));
+        std::cout << array_cvb->item_offset(1 + i) << " ++++++++" << 1 + i << std::endl;
```
Comment (yangzhg) [inline be/test/olap/column_vector_test.cpp:97]: remove this

File: be/src/olap/column_vector.cpp [hunk 2]
```diff
@@ -173,31 +186,31 @@ Status ArrayColumnVectorBatch::resize(size_t new_cap) {
...
 }
 
-void ArrayColumnVectorBatch::put_item_ordinal(segment_v2::ordinal_t* ordinals, size_t start_idx,
-                                              size_t size) {
-    size_t first_offset = _item_offsets[start_idx];
-    segment_v2::ordinal_t first_ordinal = ordinals[0];
-    size_t i = 0;
-    while (++i < size) {
-        _item_offsets[start_idx + i] = first_offset + (ordinals[i] - first_ordinal);
+void ArrayColumnVectorBatch::change_sizes_to_offsets(size_t start_idx, size_t size) {
```
Comment (yangzhg) [inline be/src/olap/column_vector.cpp:194]: get_offset maybe better

File: be/src/olap/column_vector.h [hunk 3]
```diff
@@ -181,14 +216,30 @@ class ArrayColumnVectorBatch : public ColumnVectorBatch {
...
+    // start_idx = 3, size = 2
+    //
+    // before change:
+    //                    start_idx
+    //                     |
+    // _offsets: 0    3    5     2     1
+    //
+    // after change:
+    // _offsets: 0    3    5     7     8
+    void change_sizes_to_offsets(size_t start_idx, size_t size);
```
Comment (yangzhg) [inline be/src/olap/column_vector.h:234]: Please reorganize this comment

File: be/src/olap/column_vector.cpp [hunk 4]
```diff
@@ -127,8 +127,22 @@ Status ColumnVectorBatch::create(size_t init_capacity, bool is_nullable, const T
                 return Status::NotSupported(
                         "When create ArrayColumnVectorBatch, `Field` is indispensable");
             }
+
+            std::unique_ptr<ColumnVectorBatch> elements;
+            auto array_type_info = reinterpret_cast<const ArrayTypeInfo*>(type_info);
+            RETURN_IF_ERROR(ColumnVectorBatch::create(init_capacity * 2, field->get_sub_field(0)->is_nullable(),
+                    array_type_info->item_type_info(), field->get_sub_field(0), &elements));
+
+            std::unique_ptr<ColumnVectorBatch> offsets;
```
Comment (yangzhg) [inline be/src/olap/column_vector.cpp:136]: array element size now not offset ?

File: be/src/olap/column_vector.cpp [hunk 5]
```diff
@@ -127,8 +127,22 @@ Status ColumnVectorBatch::create(size_t init_capacity, bool is_nullable, const T
...
+            TypeInfo* bigint_type_info = get_scalar_type_info(FieldType::OLAP_FIELD_TYPE_UNSIGNED_BIGINT);
+            RETURN_IF_ERROR(ColumnVectorBatch::create(init_capacity + 1, false,
+                                      bigint_type_info, nullptr, &offsets));
+
             std::unique_ptr<ColumnVectorBatch> local(
```
Comment (yangzhg) [inline be/src/olap/column_vector.cpp:141]: what does local mean

**Review by yangzhg — APPROVED** [review]
LGTM

File: be/src/olap/column_vector.cpp (hunk 5 above)
Comment (liutang123) [reply be/src/olap/column_vector.cpp:141]: It means a local variable.

**Pull request merged**